core/src/main/scala/org/http4s/internal/package.scala

```diff
@@ -124,19 +129,32 @@ package object internal {
     }
 
   // Adapted from https://github.com/typelevel/cats-effect/issues/160#issue-306054982
-  private[http4s] def fromCompletableFuture[F[_], A](fcf: F[CompletableFuture[A]])(
-      implicit F: Concurrent[F]): F[A] =
-    fcf.flatMap { cf =>
-      F.cancelable { cb =>
-        cf.handle[Unit](new BiFunction[A, Throwable, Unit] {
-          override def apply(result: A, err: Throwable): Unit = err match {
-            case null => cb(Right(result))
-            case _: CancellationException => ()
-            case ex: CompletionException if ex.getCause ne null => cb(Left(ex.getCause))
-            case ex => cb(Left(ex))
+  private[http4s] def fromCompletionStage[F[_], CF[x] <: CompletionStage[x], A](
+      fcs: F[CF[A]])(implicit F: Async[F], CS: ContextShift[F]): F[A] =
+    fcs.flatMap { cs =>
+      F.async[A] { cb =>
+          cs.handle[Unit] { (result, err) =>
+            err match {
+              case null => cb(Right(result))
+              case _: CancellationException => ()
```

Comment:
I am not really sure if it is correct to ignore `CancellationException`s, specially as we no longer "cancel" on our own.

Reply:
It still seems reasonable to me: if a cats effect is canceled, it doesn't terminate.  I'm not sure that raising a Java error makes sense.